harvester/harvester
# Add resource overcommit feature

**Problem:**
Add overcommit setting and mutating/validating admission webhook for CPU, memory and storage

**Solution:**

Introduce a new harvester setting `overcommit-config` with three keys. Each key controls different resources overcommit percentage:

- `cpu`: controls VM's request/limit on CPU via mutating admisson webhook. Default to 200%.
- `memory`: controls VM's request/limit on memory via mutating admisson webhook. Default to 200%.
- `storage`: proxies to longhorn setting resource `storage-over-provisioning-percentage`. Default to 200%.

A mutating admission webhook  is added because KubeVirt only applies that its own overcommit feature on VM without any specified `resources.limits` or `resources.requests`. Harvester here handles the case which only specifies `resources.requests`. In that case, the mutating webhook simply sets the requests as limits and use the limits to calculate overcommited resources usage as the new requested values.

**Related Issue:**
 #1429 

**Test plan:**
1. Build the new harvester and run it in cluster
1. `kubectl apply -f` the content below
    ```yaml
    apiVersion: harvesterhci.io/v1beta1
    kind: Setting
    metadata:
      name: overcommit-config
    status: {}
    value: '{"cpu":300,"memory":300,"storage":300}'
    ```
1. Head to `Advanced > Settings` and change ui-index to `https://weihanglo-harv-dashboard.s3.us-east-2.amazonaws.com/dashboard/index.html`
1. Create a VM with more core than your host machine has. It should succeed.
1. Check the pod of the VM. Propably named `virt-launcher-xxxx` in default namespace. It's CPU/memory request should be lower than what you previously set during VM creation.
